File: ruby/SpaceCity.rb
#encoding:utf-8

require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'

module Deepspace
  class SpaceCity < SpaceStation
    def initialize(base_, rest_)
      super(base_)
      @base=base_
      @collaborators=rest_
    end

    def collaborators
      return @collaborators
    end

    def fire
      fire=super
      @collaborators.each do |station|
        fire += station.fire
      end

      return fire
    end

    def protection
      protection=super
      @collaborators.each do |station|
        protection+=station.protection
      end
      return protection
    end

    def setLoot(loot)
      super(loot)
      return Transformation::NOTRANSFORM
    end

    def to_s
      getUIversion().to_s
    end
    def getUIversion
      return SpaceCityToUI.new(self)
    end
  end
end

File: Examen SpaceFighter/src/deepspace/SpaceCity.java
package deepspace;

import java.util.ArrayList;

public class SpaceCity extends SpaceStation{
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;
    public SpaceCity (SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        this.base = base;
        collaborators = new ArrayList<>(rest);
    }

    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    @Override
    public float fire(){
        float fire = super.fire();
        for(SpaceStation s : collaborators ){
            fire += s.fire();
        }
        return fire;
    }
       
    @Override
    public float protection(){
        float protection = super.protection();
        for( SpaceStation s: collaborators){
            protection += s.protection();
        }
        return protection;
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        return(Transformation.NOTRANSFORM);
    }

    @Override
    public SpaceCityToUI getUIversion() {
        return new SpaceCityToUI(this);
    }
}